server/sonar-server/src/main/java/org/sonar/server/computation/scm/DbScmInfo.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * SonarQube, open source software quality management tool.
+ * Copyright (C) 2008-2014 SonarSource
+ * mailto:contact AT sonarsource DOT com
+ *
+ * SonarQube is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.computation.scm;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.Immutable;
+import org.sonar.db.protobuf.DbFileSources;
+import org.sonar.server.computation.component.Component;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Predicates.notNull;
+import static com.google.common.collect.FluentIterable.from;
+import static com.google.common.collect.Iterables.isEmpty;
+import static java.lang.String.format;
+
+/**
+ * ScmInfo implementation based on the lines stored in DB
+ */
+@Immutable
+class DbScmInfo implements ScmInfo {
+
+  private ScmInfo delegate;
+
+  private DbScmInfo(ScmInfo delegate) {
+    this.delegate = delegate;
+  }
+
+  static Optional<ScmInfo> create(Component component, Iterable<DbFileSources.Line> lines) {
+    LineToChangeset lineToChangeset = new LineToChangeset();
+    Iterable<Changeset> lineChangesets = from(lines)
```

Comment:
warning, you should create a `List` here, otherwise iteration and transformation of `lines` are evaluated twice (in `isEmpty` and in `DbScmInfo` constructor